pkg/auth/authorizer/union/union.go

```diff
@@ -30,16 +30,25 @@ func New(authorizationHandlers ...authorizer.Authorizer) authorizer.Authorizer {
 }
 
 // Authorizes against a chain of authorizer.Authorizer objects and returns nil if successful and returns error if unsuccessful
-func (authzHandler unionAuthzHandler) Authorize(a authorizer.Attributes) error {
-	var errlist []error
+func (authzHandler unionAuthzHandler) Authorize(a authorizer.Attributes) (bool, string, error) {
+	var (
+		errlist []error
+		reason  string
+	)
 	for _, currAuthzHandler := range authzHandler {
-		err := currAuthzHandler.Authorize(a)
+		authorized, currentReason, err := currAuthzHandler.Authorize(a)
+		if currentReason != "" {
```

Comment:
keep a list of reasons, append to it in unauthorized cases (and also return the reason in authorized cases)
```
authorized, reason, err := currAuthzHandler.Authorize(a)
if err != nil {
  errlist = append(errlist, err)
  continue
}
if !authorized {
  if reason != "" {
    reasonlist = append(reasonlist, reason)
  }
  continue
}
return true, reason, nil
```